File: CSE_3206(SWE)/Design_Patterns_CPP/Decorator_Method/CoffeeShop.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>

using namespace std;

// Component Interface
class Coffee {
public:
    virtual string getDescription() const = 0;
    virtual double getCost() const = 0;
    virtual ~Coffee() = default;
};

// Concrete Component
class SimpleCoffee : public Coffee {
public:
    string getDescription() const override {
        return "Simple Coffee";
    }

    double getCost() const override {
        return 5.0;
    }
};

// Abstract Decorator
class CoffeeDecorator : public Coffee {
protected:
    shared_ptr<Coffee> coffee;

public:
    CoffeeDecorator(shared_ptr<Coffee> coffee) : coffee(std::move(coffee)) {}

    string getDescription() const override {
        return coffee->getDescription();
    }

    double getCost() const override {
        return coffee->getCost();
    }
};

// Concrete Decorators
class MilkDecorator : public CoffeeDecorator {
public:
    MilkDecorator(shared_ptr<Coffee> coffee) : CoffeeDecorator(std::move(coffee)) {}

    string getDescription() const override {
        return coffee->getDescription() + ", Milk";
    }

    double getCost() const override {
        return coffee->getCost() + 1.5;
    }
};

class SugarDecorator : public CoffeeDecorator {
public:
    SugarDecorator(shared_ptr<Coffee> coffee) : CoffeeDecorator(std::move(coffee)) {}

    string getDescription() const override {
        return coffee->getDescription() + ", Sugar";
    }

    double getCost() const override {
        return coffee->getCost() + 0.5;
    }
};

class WhippedCreamDecorator : public CoffeeDecorator {
public:
    WhippedCreamDecorator(shared_ptr<Coffee> coffee) : CoffeeDecorator(std::move(coffee)) {}

    string getDescription() const override {
        return coffee->getDescription() + ", Whipped Cream";
    }

    double getCost() const override {
        return coffee->getCost() + 2.0;
    }
};

// Main method for testing
int main(int argc, char** argv) {
    cout << "Testing Decorator Pattern Example..." << endl;

    // Run tests
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

// Google Test Cases
TEST(CoffeeShopTest, SimpleCoffeeTest) {
    shared_ptr<Coffee> coffee = make_shared<SimpleCoffee>();

    EXPECT_EQ(coffee->getDescription(), "Simple Coffee");
    EXPECT_DOUBLE_EQ(coffee->getCost(), 5.0);
}

TEST(CoffeeShopTest, CoffeeWithMilkTest) {
    shared_ptr<Coffee> coffee = make_shared<SimpleCoffee>();
    coffee = make_shared<MilkDecorator>(coffee);

    EXPECT_EQ(coffee->getDescription(), "Simple Coffee, Milk");
    EXPECT_DOUBLE_EQ(coffee->getCost(), 6.5);
}

TEST(CoffeeShopTest, CoffeeWithMilkAndSugarTest) {
    shared_ptr<Coffee> coffee = make_shared<SimpleCoffee>();
    coffee = make_shared<MilkDecorator>(coffee);
    coffee = make_shared<SugarDecorator>(coffee);

    EXPECT_EQ(coffee->getDescription(), "Simple Coffee, Milk, Sugar");
    EXPECT_DOUBLE_EQ(coffee->getCost(), 7.0);
}

TEST(CoffeeShopTest, CoffeeWithAllDecoratorsTest) {
    shared_ptr<Coffee> coffee = make_shared<SimpleCoffee>();
    coffee = make_shared<MilkDecorator>(coffee);
    coffee = make_shared<SugarDecorator>(coffee);
    coffee = make_shared<WhippedCreamDecorator>(coffee);

    EXPECT_EQ(coffee->getDescription(), "Simple Coffee, Milk, Sugar, Whipped Cream");
    EXPECT_DOUBLE_EQ(coffee->getCost(), 9.0);
}

File: CSE_3206(SWE)/Design_Pattern_Codes_Java/Decorator_Method/CoffeeShop.java
// Component Interface
interface Coffee {
    String getDescription();
    double getCost();
}

// ConcreteComponent
class SimpleCoffee implements Coffee {
    @Override
    public String getDescription() {
        return "Simple Coffee";
    }

    @Override
    public double getCost() {
        return 5.0;
    }
}

// Abstract Decorator
abstract class CoffeeDecorator implements Coffee {
    protected Coffee coffee;

    public CoffeeDecorator(Coffee coffee) {
        this.coffee = coffee;
    }

    @Override
    public String getDescription() {
        return coffee.getDescription();
    }

    @Override
    public double getCost() {
        return coffee.getCost();
    }
}

// Concrete Decorators
class MilkDecorator extends CoffeeDecorator {
    public MilkDecorator(Coffee coffee) {
        super(coffee);
    }

    @Override
    public String getDescription() {
        return coffee.getDescription() + ", Milk";
    }

    @Override
    public double getCost() {
        return coffee.getCost() + 1.5;
    }
}

class SugarDecorator extends CoffeeDecorator {
    public SugarDecorator(Coffee coffee) {
        super(coffee);
    }

    @Override
    public String getDescription() {
        return coffee.getDescription() + ", Sugar";
    }

    @Override
    public double getCost() {
        return coffee.getCost() + 0.5;
    }
}

class WhippedCreamDecorator extends CoffeeDecorator {
    public WhippedCreamDecorator(Coffee coffee) {
        super(coffee);
    }

    @Override
    public String getDescription() {
        return coffee.getDescription() + ", Whipped Cream";
    }

    @Override
    public double getCost() {
        return coffee.getCost() + 2.0;
    }
}

public class CoffeeShop {
    public static void main(String[] args) {
        // Basic Coffee
        Coffee coffee = new SimpleCoffee();
        System.out.println(coffee.getDescription() + " -> $" + coffee.getCost());

        // Add Milk
        coffee = new MilkDecorator(coffee);
        System.out.println(coffee.getDescription() + " -> $" + coffee.getCost());

        // Add Sugar
        coffee = new SugarDecorator(coffee);
        System.out.println(coffee.getDescription() + " -> $" + coffee.getCost());

        // Add Whipped Cream
        coffee = new WhippedCreamDecorator(coffee);
        System.out.println(coffee.getDescription() + " -> $" + coffee.getCost());
    }
}
